borg/crypto.pyx

```diff
@@ -63,93 +74,222 @@ def num_aes_blocks(int length):
+            rc = EVP_EncryptUpdate(&self.ctx, odata+offset, &olen, <const unsigned char*> idata.buf, ilen)
+            if not rc:
+                raise Exception('EVP_EncryptUpdate failed')
+            offset += olen
+            rc = EVP_EncryptFinal_ex(&self.ctx, odata+offset, &olen)
+            if not rc:
+                raise Exception('EVP_EncryptFinal_ex failed')
+            offset += olen
+            rc = HMAC(EVP_sha256(), self.mac_key, 32, odata+32, offset-32, odata, NULL)
+            if rc != odata:
+                raise Exception('HMAC failed')
+            return odata[:offset]
+        finally:
+            free(odata)
+            PyBuffer_Release(&idata)
+
+    def decrypt(self, data):
+        cdef Py_buffer idata = ro_buffer(data)
+        cdef int ilen = len(data)
+        cdef unsigned char *odata = <unsigned char *>malloc(ilen + 16)
+        cdef int olen
+        cdef int offset
+        cdef unsigned char *hmac
+        cdef unsigned char hmac_buf[32]
+        if not odata:
+            raise MemoryError
+        try:
+            hmac = HMAC(EVP_sha256(), self.mac_key, 32, <const unsigned char *> idata.buf+32, ilen-32, hmac_buf, NULL)
+            if hmac != hmac_buf:
+                raise Exception('HMAC failed')
+            if CRYPTO_memcmp(hmac, idata.buf, 32):
+                raise Exception('Authentication failed')
+            iv = self.fetch_iv(<unsigned char *> idata.buf)
+            if not EVP_DecryptInit_ex(&self.ctx, EVP_aes_256_ctr(), NULL, self.enc_key, iv):
                 raise Exception('EVP_DecryptInit_ex failed')
+            offset = 0
+            rc = EVP_DecryptUpdate(&self.ctx, odata+offset, &olen, <const unsigned char*> idata.buf+40, ilen-40)
+            if not rc:
+                raise Exception('EVP_DecryptUpdate failed')
+            offset += olen
+            rc = EVP_DecryptFinal_ex(&self.ctx, odata+offset, &olen)
+            if rc <= 0:
+                # this error check is very important for modes with padding or
+                # authentication. for them, a failure here means corrupted data.
+                # CTR mode does not use padding nor authentication.
+                raise Exception('EVP_DecryptFinal_ex failed')
+            offset += olen
+            return odata[:offset]
+        finally:
+            free(odata)
+            PyBuffer_Release(&idata)
+
+    cdef fetch_iv(self, unsigned char * idata):
+        # fetch lower 8 bytes of iv and add upper 8 zero bytes
+        return b"\0" * 8 + idata[32:40]
+
+    cdef store_iv(self, unsigned char * odata, unsigned char * iv):
+        # store only lower 8 bytes, upper 8 bytes are assumed to be 0
+        cdef int i
+        for i in range(8):  # XXX easier way?
+            odata[32+i] = iv[8+i]
+
+
+cdef class AES_GCM_256_GMAC:
+    # Layout: GMAC 16 + IV 16 + CT (borg 1.2)
+    # additionally, each chunk starts with a type byte,
+    # which is not passed to this code or added by this code.
```

Comment:
well, it is the default. but recommended / why not more?

i mainly wanted to avoid the byte-frickling as seen in the legacy class.